### PR TITLE
breaking(gatsby-plugin-sitemap): vNext rewrite

### LINKED ISSUE #22757: RFC(plugin-sitemap): What would you like to see fixed or changed?
## Summary
I recently did some work on `gatsby-plugin-sitemap` to resolve some edge case issues(#20692) caused by assumptions made in its code. This work fixed some issues and revealed some more. I made a note in the PR(#21948) that it might be time for a rewrite with some API changes to get all this smoothed out and to avoid some serious "jank" in the code. 

## Proposal
Basically, I have time since the world is on pause and might just get this done. 

To start this process, I wanted to see if anyone out there had any suggestions on **features to add** or **bugs to resolved** in a new version(I'm expecting their to be breaking changes currently, but we'll see what happens)?

I'll do some thinking and design work for the API in the coming days and add what I'm thinking and incorporate any feedback I get. Thanks for any input!

### PR DESCRIPTION
**Pull request by moonmeister**
## Description

Re-write of `gatsby-plugin-sitemap` to fix the spaghetti and improve the API. Mostly follows #22757, has some slight deviations where the spec needed more or i realized there was a better way. I'll make more notes as it gets closer to being ready.

TODO: 

- [x] more complete Testing before release
- [ ] Add activity bars for sitemap
- [x] remove logic for disabled options schema validation now that it's shipped
- [x] allow serialize(or other functions) to be async
- [x] improve page filtering performance

## Related Issues

closes #22757

### CODE REVIEW

**Comment by moonmeister** [comment]
Okay, update. I've rewritten the code. I've updated the docs. I've "fixed" the tests. The tests should be passing, but I left them mostly the same. I did delete a couple tests that were no longer needed. 

Next steps:
1. If someone (or multiple people) could do code review that'd be great. 
2. Once the code review is done (or before), I'd like to see this published under a `next` tag so I and others can test it in the wild (like #26707). 
3. If all goes well we can publish, or maybe we can write some more tests first but that can be determined later.

File: packages/gatsby-plugin-sitemap/README.md
Comment (moonmeister) [inline packages/gatsby-plugin-sitemap/README.md:49]:
```suggestion
You really want to customize this plugin config to include an accurate `lastmod` date. Checkout the [example](#example) for an example of how to do this.
```

**Comment by moonmeister** [comment]
Oops, I forgot to fix the tests in `internals.js` I'll do that soon. I would appreciate review on the rest in the mean time.

**Review by LekoArts — CHANGES_REQUESTED** [review]
Some first preliminary docs comments :)

File: packages/gatsby-plugin-sitemap/README.md
Comment (moonmeister) [reply packages/gatsby-plugin-sitemap/README.md:44]: Agreed.

File: packages/gatsby-plugin-sitemap/README.md
Comment (moonmeister) [reply packages/gatsby-plugin-sitemap/README.md:132]: because I wasn't about to try and figure out and remember how `## serialize ⇒ <code>object</code>` converted to an id. It was easier to create anchors that just had the function name to easily link to. If there is an A11y or other concern I'm unaware of we can find a better way or just do it the hard way.

**Comment by moonmeister** [comment]
Tests should all be passing once the latest run complete. Should be ready for more complete reviews. Thanks.

**Comment by wardpeet** [comment]
i'll publish an @next version so people can try it out.

File: packages/gatsby-plugin-sitemap/src/gatsby-node.js
Comment (wardpeet) [inline packages/gatsby-plugin-sitemap/src/gatsby-node.js:29]: This will cause many sideEffects and might lead to weird bug reports.
```suggestion
```

File: packages/gatsby-plugin-sitemap/src/gatsby-node.js
Comment (wardpeet) [inline packages/gatsby-plugin-sitemap/src/gatsby-node.js:31]:
```suggestion
  // Schema was already validated in preInit but we use joi to get our pluginOptions with default options.
  const {
```

File: packages/gatsby-plugin-sitemap/src/gatsby-node.js
Comment (wardpeet) [inline packages/gatsby-plugin-sitemap/src/gatsby-node.js:48]: Maybe add a new line?
```suggestion
    `${reporterPrefix} Query Results:\n${JSON.stringify(queryRecords, null, 2)}`
```

File: packages/gatsby-plugin-sitemap/src/gatsby-node.js
Comment (wardpeet) [inline packages/gatsby-plugin-sitemap/src/gatsby-node.js:51]: Should we make it async, so users can do whatever? I don't see an immediate use case but 🤷

File: packages/gatsby-plugin-sitemap/src/__tests__/gatsby-node.js
Comment (wardpeet) [inline packages/gatsby-plugin-sitemap/src/__tests__/gatsby-node.js:18]: Best to not show errors in console during tests
```suggestion
const reporter = {
  verbose: jest.fn(),
  panic: jest.fn(),
}
```

File: packages/gatsby-plugin-sitemap/src/__tests__/gatsby-node.js
Comment (wardpeet) [inline packages/gatsby-plugin-sitemap/src/__tests__/gatsby-node.js:9]: Shouldn't we test onPreInit too? Making sure joi fires the right errors

File: packages/gatsby-plugin-sitemap/src/__tests__/gatsby-node.js
Comment (wardpeet) [inline packages/gatsby-plugin-sitemap/src/__tests__/gatsby-node.js:26]:
```suggestion
describe(`gatsby-plugin-sitemap Node API`, () => {
```

File: packages/gatsby-plugin-sitemap/src/__tests__/gatsby-node.js
Comment (wardpeet) [inline packages/gatsby-plugin-sitemap/src/__tests__/gatsby-node.js:27]:
```suggestion
  it(`should succeed with default options`, async () => {
```

File: packages/gatsby-plugin-sitemap/src/__tests__/gatsby-node.js
Comment (wardpeet) [inline packages/gatsby-plugin-sitemap/src/__tests__/gatsby-node.js:57]:
```suggestion
  it(`should accept a custom query`, async () => {
```

File: packages/gatsby-plugin-sitemap/src/__tests__/gatsby-node.js
Comment (wardpeet) [inline packages/gatsby-plugin-sitemap/src/__tests__/gatsby-node.js:114]: I don't think we need this describe block, we can move all of it to an it clause.

File: packages/gatsby-plugin-sitemap/src/__tests__/gatsby-node.js
Comment (wardpeet) [inline packages/gatsby-plugin-sitemap/src/__tests__/gatsby-node.js:1]: Can you add tests for all functionalities like filters

File: packages/gatsby-plugin-sitemap/src/__tests__/gatsby-ssr.js
Comment (wardpeet) [inline packages/gatsby-plugin-sitemap/src/__tests__/gatsby-ssr.js:5]:
```suggestion
describe(`gatsby-plugin-sitemap SSR API`, () => {
```

File: packages/gatsby-plugin-sitemap/src/__tests__/gatsby-ssr.js
Comment (wardpeet) [inline packages/gatsby-plugin-sitemap/src/__tests__/gatsby-ssr.js:14]:
```suggestion
  it(`should create a Link if createLinkInHead is true`, async () => {
```

File: packages/gatsby-plugin-sitemap/src/__tests__/gatsby-ssr.js
Comment (wardpeet) [inline packages/gatsby-plugin-sitemap/src/__tests__/gatsby-ssr.js:30]:
```suggestion
  it(`should not create Link if createLinkInHead is false`, async () => {
```

File: packages/gatsby-plugin-sitemap/src/options-validation.js
Comment (wardpeet) [inline packages/gatsby-plugin-sitemap/src/options-validation.js:42]: This is cool!!!

File: packages/gatsby-plugin-sitemap/src/internals.js
Comment (wardpeet) [inline packages/gatsby-plugin-sitemap/src/internals.js:20]: What does this todo mean?

File: packages/gatsby-plugin-sitemap/src/internals.js
Comment (wardpeet) [inline packages/gatsby-plugin-sitemap/src/internals.js:45]: What happens if page.path is null? should we throw / warn the user?
```suggestion
  return page?.path
```

File: packages/gatsby-plugin-sitemap/src/internals.js
Comment (wardpeet) [inline packages/gatsby-plugin-sitemap/src/internals.js:1]: It's better to pass it into your functions
```suggestion
```

File: packages/gatsby-plugin-sitemap/src/gatsby-node.js
Comment (wardpeet) [inline packages/gatsby-plugin-sitemap/src/gatsby-node.js:118]: Can we extract this into a sub-function so the flow of how a sitemap is generated is easier to understand?

You probably want to combine 2 filters int 1 as this might become slow when you have 100k pages.

**Review by wardpeet — COMMENTED** [review]
Nice work @moonmeister. This is a big step up from what we had before. Our users will love it! ❤️ 

Would it be useful if we add a how-to upgrade section in the readme file?

I added some minor comments, feel free to disagree.

**Comment by wardpeet** [comment]
Published canary, you can install it with `gatsby-plugin-sitemap@next`

File: packages/gatsby-plugin-sitemap/src/__tests__/gatsby-node.js
Comment (moonmeister) [reply packages/gatsby-plugin-sitemap/src/__tests__/gatsby-node.js:18]: Oops, I think this was leftover from debugging, good catch.

File: packages/gatsby-plugin-sitemap/src/internals.js
Comment (moonmeister) [reply packages/gatsby-plugin-sitemap/src/internals.js:20]: This has been put in all over Gatsby where we do this janky asset and path prefix handling. I didn't add it but I'm happy to make it more descriptive.

File: packages/gatsby-plugin-sitemap/src/gatsby-node.js
Comment (moonmeister) [reply packages/gatsby-plugin-sitemap/src/gatsby-node.js:118]: Yes I can break this out, good call. The reason I specifically did two filters was so folks could easily write a custom filter without breaking and having to reimplement  the default filter. I'll do some thinking on how to make that more performant.

File: packages/gatsby-plugin-sitemap/src/internals.js
Comment (wardpeet) [reply packages/gatsby-plugin-sitemap/src/internals.js:20]: please do cause I forgot about it :joy:

File: packages/gatsby-plugin-sitemap/src/gatsby-node.js
Comment (wardpeet) [reply packages/gatsby-plugin-sitemap/src/gatsby-node.js:118]: I assumed it was for that reason but we can still do it in one filter loop, no?

Comment (moonmeister) [reply packages/gatsby-plugin-sitemap/src/gatsby-node.js:118]: Probably. 🤪

File: packages/gatsby-plugin-sitemap/src/gatsby-node.js
Comment (moonmeister) [reply packages/gatsby-plugin-sitemap/src/gatsby-node.js:52]: I'll await `resolveSiteUrl` as well.

File: packages/gatsby-plugin-sitemap/src/gatsby-node.js
Comment (moonmeister) [inline packages/gatsby-plugin-sitemap/src/gatsby-node.js:57]: This is here to handle both sync and async functions while also catching errors. If feels messy but I can't think of a better way. I didn't like try/catch cause I didn't want to have to handle scope. Any suggestions or is this the way?

File: packages/gatsby-plugin-sitemap/src/gatsby-node.js
Comment (moonmeister) [reply packages/gatsby-plugin-sitemap/src/gatsby-node.js]: Okay, I've moved this code out, moved it to a single filter and added some logic to reduce filter checks for the default page filters. More could probably be done but this will be much better.

File: packages/gatsby-plugin-sitemap/src/__tests__/gatsby-node.js
Comment (moonmeister) [reply packages/gatsby-plugin-sitemap/src/__tests__/gatsby-node.js:1]: The filters are being tested in the `internals.js` tests (or will be)...what tests do you think should be here?

File: packages/gatsby-plugin-sitemap/src/__tests__/gatsby-node.js
Comment (moonmeister) [reply packages/gatsby-plugin-sitemap/src/__tests__/gatsby-node.js:9]: Doing this kinda feels like we'd be testing Joi...though I suppose it'd also be testing our config and Joi objects...I'll work on that.

**Comment by moonmeister** [comment]
@wardpeet Instead of having to run the schema Validation multiple times to handle early validation and defaults, what if we used `store` to store the results from `preInit` to be used by SSR and other Node APIs? I looked into this but the docs suggest not to use store if I don't have to...is this a good use of `store` or should are we good as is? Thanks

**Comment by moonmeister** [comment]
The `e2e_tests_path-prefix` e2e test was failing due to using node `v10.16` and `Readable.Read` was introduced in `v10.17`. That docker image wasn't available so I updated the e2e runner to `v12.18`. If we need to keep e2e on `v10` We'll have to open a PR at https://github.com/cypress-io/cypress-docker-images/ to get a current `v10` release image available.

**Comment by moonmeister** [comment]
@wardpeet @LekoArts Alright, lots of tests have been written issues fixed. Should be ready for a review and response to my various questions ☝️. Thanks

File: packages/gatsby-plugin-sitemap/src/gatsby-node.js
Comment (wardpeet) [inline packages/gatsby-plugin-sitemap/src/gatsby-node.js:43]: This is a little bit cleaner with the same result
```suggestion
  const allPages = await Promise.resolve(resolvePages(queryRecords)).catch(err =>
    reporter.panic(`${ReporterPrefix} Error resolving Pages`, err)
  )
```

File: packages/gatsby-plugin-sitemap/src/gatsby-node.js
Comment (wardpeet) [inline packages/gatsby-plugin-sitemap/src/gatsby-node.js:48]:
```suggestion
  const siteUrl = await Promise.resolve(resolveSiteUrl(queryRecords)).catch(err =>
    reporter.panic(`${ReporterPrefix} Error resolving Site URL`, err)
  )
```

File: packages/gatsby-plugin-sitemap/src/gatsby-node.js
Comment (wardpeet) [inline packages/gatsby-plugin-sitemap/src/gatsby-node.js:83]:
```suggestion
  const serializedPages [];
  filteredPages.forEach(page => {
    try {
      const { url, ...rest } = serialize(page, { resolvePagePath })
      serializedPages.push({ url: prefixPath({ url, siteUrl, pathPrefix }), ...rest })
    } catch (err) {
      reporter.panic(`${ReporterPrefix} Error serializing pages`, err)
    }
```

File: packages/gatsby-plugin-sitemap/src/gatsby-node.js
Comment (wardpeet) [inline packages/gatsby-plugin-sitemap/src/gatsby-node.js:60]: I think this is a requisite from the past
```suggestion
```

File: packages/gatsby-plugin-sitemap/src/internals.js
Comment (wardpeet) [inline packages/gatsby-plugin-sitemap/src/internals.js:177]: With an explicit return we can remove the eslint disable comment
```suggestion
      }
      
      return false
    })
```

File: packages/gatsby-plugin-sitemap/src/internals.js
Comment (wardpeet) [inline packages/gatsby-plugin-sitemap/src/internals.js:192]: can we add return false to the bottom so we can drop this line

**Review by wardpeet — DISMISSED** [review]
Some small nits, I think afterwards we can merge this in. Thank you @moonmeister

**Comment by moonmeister** [comment]
> 
> 
> The `e2e_tests_path-prefix` e2e test was failing due to using node `v10.16` and `Readable.Read` was introduced in `v10.17`. That docker image wasn't available so I updated the e2e runner to `v12.18`. If we need to keep e2e on `v10` We'll have to open a PR at https://github.com/cypress-io/cypress-docker-images/ to get a current `v10` release image available.

@wardpeet can you respond to this?

**Comment by moonmeister** [comment]
@mxstbr Do you have an idea of how long till the `pluginOptionsSchema` API will be enabled by default? I'd like to use it for this plugin update. Thanks.

**Comment by moonmeister** [comment]
@wardpeet Since I'm not working at Gatsby an more I've lost write permissions and can no longer update this branch. I have a use merge and fixes to push but can't...Can I get write access again till this is merged? if not we'll need to find another way.

**Comment by wardpeet** [comment]
I've given you permissions, i'll think about the cypress one.

**Comment by moonmeister** [comment]
@wardpeet Can we get this across the finish line?

**Comment by moonmeister** [comment]
@wardpeet This is ready to go from my perspective. Just need to resolve that e2e test issue runner question(https://github.com/gatsbyjs/gatsby/pull/25670#issuecomment-695895824).

**Comment by moonmeister** [comment]
https://github.com/gatsbyjs/gatsby/pull/28742 just updated the e2e runner so that's been resolved. This is ready to merge assuming all is green.

**Comment by wardpeet** [comment]
Thanks for waiting @moonmeister I'll merge it this week, have to do a couple of tests but code looks awesome!

**Comment by moonmeister** [comment]
@wardpeet Don't mean to push, take your time for any final release, but if these tests pass could you at-least update the @next version we published previously? Thanks.

**Comment by moonmeister** [comment]
Changes look good to me ward. Thanks.

**Comment by moonmeister** [comment]
As shown by all the folks over at #29810 and #22757 this update is in demand, can we get this merged? thanks!

**Comment by moonmeister** [comment]
@KyleAMathews what can we do to get this merged?

**Review by wardpeet — APPROVED** [review]
Thanks for waiting, I was on vacation. I'll merge it in and publish a @next version for it!

Thanks a bunch @moonmeister

**Comment by moonmeister** [comment]
Thanks @wardpeet!

**Comment by t2ca** [comment]
Just want to let you know that when I tried this with just the minimal configuration, i got the following error

```
"gatsby-plugin-sitemap" threw an error while running the onPostBuild lifecycle:

Body must be a string. Received: undefined.

  38 |             output = _ref2.output, entryLimit = _ref2.entryLimit, query = _ref2.query, excludes =
_ref2.excludes, resolveSiteUrl = _ref2.resolveSiteUrl, resolvePagePath = _ref2.resolvePagePath, resolvePages =
 _ref2.resolvePages, filterPages = _ref2.filterPages, serialize = _ref2.serialize;
  39 |             _context.next = 4;
> 40 |             return graphql(query);
     |                    ^
  41 |
  42 |           case 4:
  43 |             _yield$graphql = _context.sent;

File: node_modules/gatsby-plugin-sitemap/gatsby-node.js:40:20
```

**Comment by pedrolamas** [comment]
As far as I can see, the plugin is currently broken...

By default, it will generate the sitemap xml files under the "/sitemap" folder, and it does that correctly, however the sitemap index file points to "<siteUrl>/sitemap-0.xml" instead of "<siteUrl>/sitemap/sitemap-0.xml" as expected.

The easy fix would be to generate the sitemap on the root as before, however the plugin is not allowing to pass an empty string to the "output" option.

This issue is currently capturing some of the issues found on the 4.0.0 release of the plugin: https://github.com/gatsbyjs/gatsby/issues/31095

**Comment by axe312ger** [comment]
Can confirm, getting the same error with v4.0.0 and v4.1.0-next.0.

```sh
"gatsby-plugin-sitemap" threw an error while running the onPostBuild lifecycle:

Body must be a string. Received: undefined.

  38 |             output = _ref2.output, entryLimit = _ref2.entryLimit, query = _ref2.query, excludes = _ref2.excludes, resolveSiteUrl = _ref2.resolveSiteUrl, resolvePagePath = _ref2.resolvePagePath, resolvePages =
_ref2.resolvePages, filterPages = _ref2.filterPages, serialize = _ref2.serialize;
  39 |             _context.next = 4;
> 40 |             return graphql(query);
     |                    ^
  41 |
  42 |           case 4:
  43 |             _yield$graphql = _context.sent;

```

**Comment by moonmeister** [comment]
Hi @axe312ger !

Sorry to hear you're running into an issue. To help us best begin debugging the underlying cause, it is incredibly helpful if you open a new issue. Please include a [minimal reproduction][reproduction] and a copy of any relevant code/config or a link to the repository. 

Thanks for using Gatsby! 💜

[reproduction]: https://gatsby.dev/reproduction

**Pull request merged**
